Proposed Draft of the Rust Core WG Charter (#11)
TBD54566975/technical-steering-committee: charters/rust-core/README.md

# Rust Core Working Group Charter <!-- omit in toc -->

**Start Date**: June 6th 2024

> **End Date**: Unknown

Reviewer: I would recommend being fairly aggressive with this date. Reason being:

1. It forces us to re-consider the scope of work and whether we're on the right track or need to make adjustments within a given timeframe
2. It lights a 🔥 under our 🍑 to get some work done
3. We can always re-charter to extend the date and/or make other changes

It's also fine to have a smaller set of deliverables in this charter, but call out future deliverables: like if there's a deliverable we know will be done in 3 months and another that will take 1+ year we can keep the second deliverable out of scope, or narrow the scope to something like a 'prototype'.

I generally think a charter should be under a year in length, probably 2 quarters or so, which gives us enough time to plan and commit, but not too much for things to dramatically shift.

Author: Appreciate the comment @decentralgabe! Going to work on adding timelines here shortly. 

Deliverables list seems long here but I think it's necessary. We could remove the Swift deliverables but that seems ill-advised because we want to showcase the power of cross-language bindings for more than one language, else we may naturally build matters proprietary to the single target language.

**DRI**: Kendall Weihe (@KendallWeihe)

- [Motivation and Background](#motivation-and-background)
- [Scope](#scope)
- [Deliverables](#deliverables)
  - [Web5](#Web5)
    - [\[Deliverable 1\]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid)
    - [\[Deliverable 2\]: Web5 Rust Implementation](#deliverable-2-web5-rust-implementation)
    - [\[Deliverable 3\]: Web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding)

>     - [\[Deliverable 4\]: Web5 Kotlin Implementation](#deliverable-4-web5-kotlin-implementation)

Reviewer 2: i dunno if these are sequential but i'd vote for web5 X implementation and tbdex X implementation to go hand in hand where X is language of choice

which means you'd want to do web5 API design and tbdex API design together

the reason for this is that tbdex heavily relies on web5 api to do web5 primitive work

Author: Good callout, I think this'll occur through DRI-ship. @nitro-neal has both Kotlin implementations, @kirahsapong has both Swift implementations, and I have both API Designs. We are split on Rust where @diehuxx has tbDEX and I have Web5, but this could be fine? @diehuxx how do you feel about being DRI over both Web5 & tbDEX Rust Implementation deliverables? If you're willing, we could conform to @jiyoontbd's comment completely, through DRI-ship if you're take on both Web5 and tbDEX.

    - [\[Deliverable 5\]: Web5 Swift Implementation](#deliverable-5-web5-swift-implementation)
  - [tbDEX](#tbdex)
    - [\[Deliverable 6\]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid)
    - [\[Deliverable 7\]: tbDEX Rust Implementation](#deliverable-7-tbdex-rust-implementation)
    - [\[Deliverable 8\]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding)
    - [\[Deliverable 9\]: tbDEX Kotlin Implementation](#deliverable-9-tbdex-kotlin-implementation)
    - [\[Deliverable 10\]: tbDEX Swift Implementation](#deliverable-10-tbdex-swift-implementation)
- [Success Criteria](#success-criteria)
- [Coordination](#coordination)
- [Communication](#communication)

## Motivation and Background

> This working group will streamline and standardize the implementations of the open standards presented by the Web5 and tbDEX Working Groups.

Reviewer 3: ```suggestion
This working group will develop a conformant, secure, and maintainable implementation of the Web5 and tbDEX specifications by creating a Rust core SDK and providing language bindings.
```

- Contain a convenient location for a historical record of decisions-made.

Rust is uniquely qualified as the solution for a cross-platform core implementation because of open source tooling availability through WASM and UniFFI. Furthermore, Rust has a broad array of open source support, and Rust has the unique feature of memory safety (relative to alternatives such as C).

## Scope

The comprehensive implementations of the Web5 and tbDEX specifications across Rust, Kotlin and Swift are within the scope of this Working Group. Alongside of each implementation, documentation comments ("doc comments"), test vectors, example usages, and CI/CD solutions (eg. security analysis, test automation, artifact version & publication, etc.) are all within scope of this Working Group. 

The Rust Core Working Group will make an intentional effort to prioritize matters of security and performance. Furthermore, with regards to API design & developer experience (DX), this Working Group will prioritize first and foremost modularity and extensibility, over DX convenience, in order to limit the liklihood of introducing breaking changes to consumers of the delivered SDKs.

> The Rust Core Working Group is not concerned with the definitions of the Web5 and tbDEX specifications, both of which are the concerns of the respective Working Group. However, the Rust Core Working Group will offer valuable concrete feedback to the Web5 and tbDEX Working Groups. Although the Web5 and tbDEX specifications are not within the scope of this Working Group, a considerable amount of cross-Working Group collaboration is to be expected (see [Coordination](#coordination) and [Communication](#communication) below).

Reviewer 3: ```suggestion
The Rust Core Working Group is not responsible for defining the Web5 and tbDEX specifications, as this falls under the purview of their respective Working Groups. However, the Rust Core Working Group will offer valuable concrete feedback to the Web5 and tbDEX Working Groups. Although the Web5 and tbDEX specifications are not within the scope of this Working Group, a considerable amount of cross-Working Group collaboration is to be expected (see [Coordination](#coordination) and [Communication](#communication) below).
```

> 
> The Rust Core Working Group is not concerned with the definitions of the web5 and tbdex specifications, both of which are the concerns of the respective Working Group. However, the Rust Core Working Group will offer valuable concrete feedback to the web5 and tbdex Working Groups. Although the web5 and tbdex specifications are not within the scope of this Working Group, a considerable amount of cross-Working Group collaboration is to be expected (see [Coordination](#coordination) and [Communication](#communication) below).
> 
> The Rust Core Working Group codifies a standard API Design ([web5 APID](#deliverable-1-web5-api-design-apid) & [tbdex APID](#deliverable-6-tbdex-api-design-apid)) which is implemented across all supported language implementations (Rust, Kotlin and Swift), but any additional API features or design characteristics are not within the scope of this Working Group. Each implementation must support the standard API Design as a baseline implementation, but does not exclude the possibility of additional features.

Reviewer: ```suggestion

The Rust Core Working Group is not concerned with the definitions of the Web5 and tbDEX specifications, both of which are the concerns of the respective Working Group. However, the Rust Core Working Group will offer valuable concrete feedback to the Web5 and tbDEX Working Groups. Although the Web5 and tbDEX specifications are not within the scope of this Working Group, a considerable amount of cross-Working Group collaboration is to be expected (see [Coordination](#coordination) and [Communication](#communication) below).

The Rust Core Working Group codifies a standard API Design ([web5 APID](#deliverable-1-web5-api-design-apid) & [tbDEX APID](#deliverable-6-tbdex-api-design-apid)) which is implemented across all supported language implementations (Rust, Kotlin and Swift), but any additional API features or design characteristics are not within the scope of this Working Group. Each implementation must support the standard API Design as a baseline implementation, but does not exclude the possibility of additional features.
```

The Rust Core Working Group will codify itself across the following GitHub repos:

> - Web5: https://github.com/TBD54566975/Web5-rs/

Reviewer 2: is the deliverable list missing web5 / tbdex wasm binding / JS implementation? forgive me, i don't know how wasm works - but you mentioned wasm above as one of the toolings to enable bindings

Author: WASM is intentionally excluded for the time being. We have plenty to do with the existing set of deliverables, so we should do what's here in full whereafter we can amend the charter to include WASM if we deem it worth including.

> - web5: https://github.com/TBD54566975/web5-rs/
>   - [[Deliverable 1]: web5 API Design (APID)](#deliverable-1-web5-api-design-apid)
>   - [[Deliverable 2]: web5 Rust Implementation](#deliverable-2-web5-rust-implementation)
>   - [[Deliverable 3]: web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding)
>   - [[Deliverable 4]: web5 Kotlin Implementation](#deliverable-4-web5-kotlin-implementation)
>   - [[Deliverable 5]: web5 Swift Implementation](#deliverable-5-web5-swift-implementation)
>   - RFCs for historical tracking of significant decisions
> - tbdex: https://github.com/TBD54566975/tbdex-rs/
>   - [[Deliverable 6]: tbdex API Design (APID)](#deliverable-6-tbdex-api-design-apid)
>   - [[Deliverable 7]: tbdex Rust Implementation](#deliverable-7-tbdex-rust-implementation)
>   - [[Deliverable 8]: tbdex UniFFI Binding](#deliverable-8-tbdex-uniffi-binding)
>   - [[Deliverable 9]: tbdex Kotlin Implementation](#deliverable-9-tbdex-kotlin-implementation)
>   - [[Deliverable 10]: tbdex Swift Implementation](#deliverable-10-tbdex-swift-implementation)
>   - RFCs for historical tracking of significant decisions

Reviewer: ```suggestion
- Web5: https://github.com/TBD54566975/web5-rs/
  - [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid)
  - [[Deliverable 2]: Web5 Rust Implementation](#deliverable-2-web5-rust-implementation)
  - [[Deliverable 3]: Web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding)
  - [[Deliverable 4]: Web5 Kotlin Implementation](#deliverable-4-web5-kotlin-implementation)
  - [[Deliverable 5]: Web5 Swift Implementation](#deliverable-5-web5-swift-implementation)
  - RFCs for historical tracking of significant decisions
- tbDEX: https://github.com/TBD54566975/tbdex-rs/
  - [[Deliverable 6]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid)
  - [[Deliverable 7]: tbDEX Rust Implementation](#deliverable-7-tbdex-rust-implementation)
  - [[Deliverable 8]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding)
  - [[Deliverable 9]: tbDEX Kotlin Implementation](#deliverable-9-tbdex-kotlin-implementation)
  - [[Deliverable 10]: tbDEX Swift Implementation](#deliverable-10-tbdex-swift-implementation)
  - RFCs for historical tracking of significant decisions
```

Author: I was hoping someone would chime in with casing 😃

## Deliverables

![Deliverables](./deliverables.png)

### web5

#### [Deliverable 1]: web5 API Design (APID)

> **Description**: The *web5 APID* is a language-independent codified API Design which serves to act as a baseline of features for implementations. The APID is primarily considered to be a deliverable for internal usage, but does not exclude the possibility of external consumption.

Reviewer: ```suggestion
**Description**: The *Web5 APID* is a language-independent codified API Design which serves to act as a baseline of features for implementations. The APID is primarily considered to be a deliverable for internal usage, but does not exclude the possibility of external consumption.
```

**DRI**: [Kendall Weihe](https://github.com/KendallWeihe).

**Scope of Work**:

- Custom DSL which is conceptually compatible with all target languages.
- Codification of the API Design using the Custom DSL.
- Example usages.

> - Doc comments inline for the API Design .

Author: ```suggestion
- Doc comments for the API Design.
```

- Test vector resources.

**Timeline/Milestones**: Unknown???

**Exit Criteria**: Comprehensive feature coverage for the web5 specification, codified in the Custom DSL, committed to source code.

#### [Deliverable 2]: web5 Rust Implementation

> **Description**: Rust implementation of the [[Deliverable 1]: web5 API Design (APID)](#deliverable-1-web5-api-design-apid), intended for public consumption.

Reviewer: ```suggestion
**Description**: Rust implementation of the [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid), intended for public consumption.
```

**DRI**: Kendall Weihe (@KendallWeihe).

**Scope of Work**:

> - Full conformance to [[Deliverable 1]: web5 API Design (APID)](#deliverable-1-web5-api-design-apid).

Reviewer: ```suggestion
- Full conformance to [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid).
```

- Comprehensive integration of test vectors (from APID).
- Comprehensive documentation comments (from APID).
- CI/CD implementations for security analysis, test automation, and crate publication.

**Timeline/Milestones**: ???

**Exit Criteria**: Full implementation of *Scope of Work* as a Rust project.

#### [Deliverable 3]: web5 UniFFI Binding

> **Description**: Rust crate which utilizes [UniFFI](https://github.com/mozilla/uniffi-rs) in conjunction with the [[Deliverable 2]: web5 Rust Implementation](#deliverable-2-web5-rust-implementation) to create Kotlin and Swift binded code, which feed into [[Deliverable 4]: web5 Kotlin Implementation](#deliverable-4-web5-kotlin-implementation) and [[Deliverable 5]: web5 Swift Implementation](#deliverable-5-web5-swift-implementation). This deliverable is intended primarily for internal purposes but does not exclude the possibility of public consumption.

Reviewer: ```suggestion
**Description**: Rust crate which utilizes [UniFFI](https://github.com/mozilla/uniffi-rs) in conjunction with the [[Deliverable 2]: Web5 Rust Implementation](#deliverable-2-web5-rust-implementation) to create Kotlin and Swift bound code, which feed into [[Deliverable 4]: web5 Kotlin Implementation](#deliverable-4-web5-kotlin-implementation) and [[Deliverable 5]: Web5 Swift Implementation](#deliverable-5-web5-swift-implementation). This deliverable is intended primarily for internal purposes but does not exclude the possibility of public consumption.
```

Author: Oooh "bound" 😆 it's the little things

**DRI**: Kendall Weihe (@KendallWeihe).

**Scope of Work**:

> - UDL file with comprehensive feature coverage from [[Deliverable 2]: web5 Rust Implementation](#deliverable-2-web5-rust-implementation).

Reviewer: ```suggestion
- UDL file with comprehensive feature coverage from [[Deliverable 2]: Web5 Rust Implementation](#deliverable-2-web5-rust-implementation).
```

- Successful binding executions for Kotlin & Swift.

**Timeline/Milestones**: ???

> **Exit Criteria**: Successful binding executions, over the comprehensive set of features from [[Deliverable 2]: web5 Rust Implementation](#deliverable-2-web5-rust-implementation), for Kotlin and Swift.
> 
> #### [Deliverable 4]: web5 Kotlin Implementation
> 
> **Description**: Kotlin implementation of web5 by wrapping [[Deliverable 3]: web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding) as to be conformant to the [[Deliverable 1]: web5 API Design (APID)](#deliverable-1-web5-api-design-apid), intended for public consumption.

Reviewer: ```suggestion
**Exit Criteria**: Successful binding executions, over the comprehensive set of features from [[Deliverable 2]: Web5 Rust Implementation](#deliverable-2-web5-rust-implementation), for Kotlin and Swift.

#### [Deliverable 4]: Web5 Kotlin Implementation

**Description**: Kotlin implementation of Web5 by wrapping [[Deliverable 3]: Web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding) as to be conformant to the [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid), intended for public consumption.
```

**DRI**: Neal Roessler (@nitro-neal).

**Scope of Work**:

- Utilization of the output of the [[Deliverable 3]: Web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding).
- Full conformance to [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid).
- Comprehensive integration of test vectors (from APID).
- Comprehensive documentation comments (from APID).
- CI/CD implementations for security analysis, test automation, and crate publication.

> **Timeline/Milestones**: ???

Reviewer 2: this date must be tied to our internal deadlines, fyi

**DRI**: [Kirah Sapong](https://github.com/kirahsapong).

**Scope of Work**:

- Utilization of the output of the [[Deliverable 3]: Web5 UniFFI Binding](#deliverable-3-web5-uniffi-binding).
- Full conformance to [[Deliverable 1]: Web5 API Design (APID)](#deliverable-1-web5-api-design-apid).
- Comprehensive integration of test vectors (from APID).
- Comprehensive documentation comments (from APID).
- CI/CD implementations for security analysis, test automation, and crate publication.

**Timeline/Milestones**: ???

**Exit Criteria**: Full implementation of *Scope of Work* as a Swift project.

### tbDEX

#### [Deliverable 6]: tbDEX API Design (APID)

**Description**: The *tbDEX APID* is a language-independent codified API Design which serves to act as a baseline of features for implementations. The APID is primarily considered to be a deliverable for internal usage, but does not exclude the possibility of external consumption.

**DRI**: [Kendall Weihe](https://github.com/KendallWeihe)

**Scope of Work**:

- Custom DSL which is conceptually compatible with all target languages.
- Codification of the API Design using the Custom DSL.
- Example usages.

> - Doc comments inline for the API Design.

Author: ```suggestion
- Doc comments for the API Design.
```

- Successful binding executions for Kotlin & Swift.

**Timeline/Milestones**: ???

**Exit Criteria**: Successful binding executions, over the comprehensive set of features from [[Deliverable 7]: tbDEX Rust Implementation](#deliverable-7-tbdex-rust-implementation), for Kotlin and Swift.

#### [Deliverable 9]: tbDEX Kotlin Implementation

**Description**: Kotlin implementation of tbDEX by wrapping [[Deliverable 8]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding) as to be conformant to the [[Deliverable 6]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid), intended for public consumption.

**DRI**: Neal Roessler (@nitro-neal).

**Scope of Work**:

- Utilization of the output of the [[Deliverable 8]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding).
- Full conformance to [[Deliverable 6]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid).
- Comprehensive integration of test vectors (from APID).
- Comprehensive documentation comments (from APID).
- CI/CD implementations for security analysis, test automation, and crate publication.

> **Timeline/Milestones**: ???

Reviewer 2: this date must be tied to our internal deadlines, fyi

Author: Yes and no. What's being proposed with the WG has a higher set of expectations than early (& narrower) use cases. AKA we can cut corners with the early use cases on things like documentation, test vectors, CI/CD, etc. Early use cases have a high-bar with respect to expectations, I don't mean to imply the inverse, because we have to establish high levels of confidence, but early use cases have a smaller footprint than the scope of work for the WG (& the deliverable specifically commented on here).

**Exit Criteria**: Full implementation of *Scope of Work* as a Kotlin project.

#### [Deliverable 10]: tbDEX Swift Implementation

**Description**: Swift implementation of tbDEX by wrapping [[Deliverable 8]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding) as to be conformant to the [[Deliverable 6]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid), intended for public consumption.

**DRI**: [Kirah Sapong](https://github.com/kirahsapong).

**Scope of Work**:

- Utilization of the output of the [[Deliverable 8]: tbDEX UniFFI Binding](#deliverable-8-tbdex-uniffi-binding).
- Full conformance to [[Deliverable 6]: tbDEX API Design (APID)](#deliverable-6-tbdex-api-design-apid).
- Comprehensive integration of test vectors (from APID).
- Comprehensive documentation comments (from APID).
- CI/CD implementations for security analysis, test automation, and crate publication.

**Timeline/Milestones**: ???

**Exit Criteria**: Full implementation of *Scope of Work* as a Swift project.

## Success Criteria

> The Rust Core Working Group will have achieved its objective if it delivers comprehensive implementations for the Web5 and tbDEX specifications, across Rust, Kotlin and Swift, as robust, extensible, performant, and secure artifacts for public consumption in highly available an reliable production environments.

Author: ```suggestion
The Rust Core Working Group will have achieved its objective if it delivers comprehensive implementations for the Web5 and tbDEX specifications, across Rust, Kotlin and Swift, as robust, extensible, performant, and secure artifacts for public consumption in highly available and reliable production environments.
```

Author: ```suggestion
The Rust Core Working Group will have achieved its objective if it delivers comprehensive implementations for the Web5 and tbDEX specifications, across Rust, Kotlin and Swift, as robust, extensible, performant, and secure artifacts for public consumption in highly available and reliable production environments; the _Exit Criteria_ for all [Deliverables](#deliverables) must be met for the WG to be considered comprehensively successful.
```

## Coordination

The Rust Core Working Group will need to coordinate regularly with the Web5 and tbDEX Working Groups as a means of consuming the specification output of said Working Groups, and providing concrete feedback to said Working Groups which arise during implementation. The Rust Core Working Group will require the Web5 and tbDEX Working Groups to produce codified specification resources which enable implementation, and will provide concrete feedback in the form of meeting attendance, and the codified resources of open source development (Issue tickets, Pull Requests, comments, & RFCs).

> In order to coordinate with the Web5 and tbDEX Working Groups, the Rust Core Working Group will elect the following individuals to represent the Rust Core Working Group within the respective Working Groups:

Author: ```suggestion
In order to coordinate with the Web5 and tbDEX Working Groups, the Rust Core Working Group elects the following individuals to represent the Rust Core Working Group within the respective Working Groups:
```

- Web5: [Neal Roessler](https://github.com/nitro-neal)
- tbDEX: [Diane Huxley](https://github.com/diehuxx)

The above individuals will be required to participate in the respective Working Group and bridge the gap between consumption of said Working Groups Deliverables, and concrete feedback.

## Communication

See [Scope](#scope) for description of GitHub repository makeup.

All productive work for the Rust Core Working Group will occur out in the open, on GitHub, for anyone to participate in.

The Rust Core Working Group will follow an RFC process, codified in the respective GitHub repository, for the matter of significant decision making, wherein Pull Requests will offer a place of communication among contributors to provide feedback.

> The Rust Core Working Group will organize weekly meetings to synchronously communicate and coordinate current affairs, and each meeting will be summarized via note taking which is codified and committed to the respective GitHub repository.

Author: Actually, we're going to codify WG meeting notes inside this repo, the TSC repo
